docs/docs-beta/docs/guides/external-systems/non-python.md

```diff
@@ -1,4 +1,141 @@
+
+Dagster Pipes follows a similar design to Unix pipes, hence the name. The `PipesSubprocessClient` is responsible for running external processes and setting up input/output files. The asset defined here is materialized using the `PipesSubprocessClient` running a Node.js file containing the `train_model` function.
+
+# Add JavaScript utility functions to access pipes
+
+The `PipesSubprocessClient` calls the child process with two environment variables defined, each containing a path to a file. One for input and one for output.
+- DAGSTER_PIPES_CONTEXT: Input context
+- DAGSTER_PIPES_MESSAGES: Output context
+
+Both environment variables are base64, zip compressed JSON objects with a "path" key. These functions decode these environment variables and access the files to hook up our JavaScript function to Dagster.
+
+```javascript
+import * as util from 'util';
+import { promises as fs } from "fs";
+import { inflate } from 'zlib';
+
+const inflateAsync = util.promisify(inflate);
+
+async function _decodeParam(value) {
+  if (!value) {
+    return null;
+  }
+  // Decode from base64 and unzip
+  const decoded = Buffer.from(value, "base64");
+  const decompressed = await inflateAsync(decoded);
+  // Deserialize to JSON
+  return JSON.parse(decompressed.toString("utf-8"));
+}
+
+async function getPipesContext() {
+  // get the env var value for where the pipes context is stored
+  const encodedPipesContextParam = process.env.DAGSTER_PIPES_CONTEXT;
+  // decove the value to get the input file path
+  const decodedPipesContextParam = await _decodeParam(encodedPipesContextParam);
+  if (!decodedPipesContextParam) {
+    return null;
+  }
+  return await fs.readFile(decodedPipesContextParam.path, "utf-8")
+    .then((data) => JSON.parse(data));
+}
+
+async function setPipesMessages(message) {
+  // get the env var value for where the pipes message is stored
+  const encodedPipesMessagesParam = process.env.DAGSTER_PIPES_MESSAGES;
+  // decode the value to get the output file path
+  const decodedPipesMessagesParam = await _decodeParam(encodedPipesMessagesParam);
+  if (!decodedPipesMessagesParam) {
+    return null;
+  }
+  const path = decodedPipesMessagesParam.path;
+  await fs.appendFile(path, JSON.stringify(message) + "\n");
+}
+```
+
+# Create an API for Dagster to invoke
+
+With the utility functions to decode the Dagster Pipes environment variables, a JavaScript function is needed to translate the Dagster Pipes context into a function evocation.
+
+```javascript
+const ALL_OPERATIONS = {
+  train_model
+}
+
+async function run_operation() {
+  const { asset_keys, extras: { operation_name, config } } = await getPipesContext()
+  if (!(operation_name in ALL_OPERATIONS)) {
+    setPipesMessages({ error: `Operation ${operation_name} not found` });
+    return;
+  }
+  const operation = ALL_OPERATIONS[operation_name];
+  const model = await operation(config);
+  await setPipesMessages({
+    method: "report_asset_materialization",
```

Comment:
I think this part is very confusing. The structure here is worth describing before introducing it

Reply:
I tried to summarize what this logic does a bit. If the feedback is that the code is too confusing then I don't think that I'll have enough time to completely rework the example. It took me all day monday and tuesday to get this far.

Reply:
i think we can land the simplest one without this whole util functions, meaning you could stop here: https://github.com/dagster-io/dagster/pull/23989#discussion_r1737281472, no context loader or message writer, and call it a day for docathon. i can pick up the later part.

Reply:
It's okay. I'll just abandon and you can pick it up later. Sorry guys.

Reply:
I think this example is still quite useful!

I'll see if we can combine it with @prha 's guide, but I don't want to lose what you produced here. Thanks @sbquinlan !